misc/utils.py

```diff
@@ -0,0 +1,155 @@
+from dataclasses import dataclass, asdict
+import datetime
+import os
+import random
+import shutil
+import subprocess
+
+from pathlib import Path
+
+import numpy as np
+import pandas as pd
+
+
+def set_seed(seed):
+    """Fix seeds for reproducable experiments."""
+    random.seed(seed)
+    np.random.seed(seed)
+    os.environ["PYTHONHASHSEED"] = str(seed)
```

Comment:
yes, to make all results fully reproducible

Reply:
Seems a bit hacky for me. Which parts of the code present undeterministic behavior because of the hash function? Maybe there is a way to redesign them

Reply:
Yes, it's a hacky approach, but I think it's the best one. At least it works (see the e2e test). The reason for fixing random seed is to make all experiments reproducible. For instance, the ICM is a stochastic process, so to force results being the same regardless the OS, time of execution and so on, I have to fix the seed. I think we also should do that in the training pipeline in order to make the training reproducible (at least I did that in my recent paper).

Reply:
Fixing random seed is one thing - I'm not a fan of handling the global numpy seed, but sure - I understand the idea behind it. 

Fixing Pythons hash seed is another thing and there is multiple issues with it:

1. Hash function is randomized on purpose, I wouldn't touch it without a very solid reason
2. It's hard for me to see where exactly could it have an impact on the experiments
3. They way you are trying to do does not even work. When you set hash seed after the interpreter is running, it is already too late. If you really want reproducible hash behavior between runs, you need to set a env variable like `PYTHONHASHSEED=0 python YOURSCRIPT.py`

Reply:
https://www.cs.cornell.edu/home/kleinber/kdd03-inf.pdf -> independent cascade model is a stochastic process. I want to produce a dataset which can be reproduced by anyone with this code. Since ICM is stochastic i have to fix the seed of a random number generator to achieve that. I think that is is a proper reason to use this function. 

FYI - I've added todo statement to check that env variable when I'll be using this function in another repositories

Reply:
I do understand that this model is stochastic, my question is a bit different tho. Have you experienced any randomness in the models behaviour that was fixed by manually setting the python hash seed?